src/keygen/keygen_commit.rs

```diff
@@ -84,27 +86,33 @@ impl KeygenDecommit {
     }
 
     #[instrument(skip_all, err(Debug))]
-    /// `sid` is a unique session identifier.
-    pub(crate) fn verify(
+    /// Verify this KeygenDecommit against a commitment and expected content.
+    fn verify(
         &self,
-        sid: &Identifier,
-        sender: &ParticipantIdentifier,
+        sid: Identifier,
+        sender: ParticipantIdentifier,
         com: &KeygenCommit,
     ) -> Result<()> {
-        let mut transcript = Transcript::new(b"KeyGenR1");
-        let decom = &mut self.clone();
-        decom.sid = *sid;
-        decom.sender = *sender;
-        transcript.append_message(b"decom", &serialize!(&decom)?);
-        let mut hash = [0u8; 32];
-        transcript.challenge_bytes(b"hashing r1", &mut hash);
-        let rebuilt_com = KeygenCommit { hash };
-
-        if rebuilt_com == *com {
-            Ok(())
-        } else {
+        // Check the commitment.
+        let rebuilt_com = self.commit()?;
+        if &rebuilt_com != com {
             error!("decommitment does not match original commitment");
-            Err(InternalError::ProtocolError(Some(*sender)))
+            return Err(InternalError::ProtocolError(Some(sender)));
+        }
+
+        // Check the session ID and sender ID.
+        if self.sid != sid {
+            error!("Incorrect session ID");
+            return Err(InternalError::ProtocolError(Some(sender)));
+        }
+        if self.sender != sender {
+            error!("Incorrect sender ID");
+            return Err(InternalError::ProtocolError(Some(sender)));
+        }
+        if self.pk.participant() != sender {
```

Comment:
Is this needed because there is not a tight binding self.pk.participant and sender?

Reply:
Well these two fields are redundant and `sender` could be removed, but if it is here it would be unclear and error-prone to not validate it. Same with `sid`.

There is a similar structure in key refresh, except `sender` is not redundant in that case.